sunkit_image/nafe.py

```diff
@@ -0,0 +1,411 @@
+  of fine structures in extreme ultraviolet images of the corona is presented
+  in this paper. The Noise Adaptive Fuzzy Equalization method is particularly
+  suited for the exceptionally high dynamic range images from the Atmospheric
+  Imaging Assembly instrument on the Solar Dynamics Observatory. This method
+  produces artifact-free images and gives significantly better results than
+  methods based on convolution or Fourier transform which are often used for
+  that purpose.}
+}
+"""
+import argparse
+import multiprocessing
+from multiprocessing import Pool
+
+import sunpy.map
+
+import skimage
+from skimage import filters
+
+import matplotlib.pyplot as plt
+import numpy as np
+import tqdm
+
+
+_PREDEF_PARAMS = {
+    # defined by original program
+    # wavelength    # gamma     # nafe
+    "4500":          (1.0,        .05),
+    "1700":          (2.2,        .10),
+    "1600":          (2.2,        .10),
+    "335":           (3.0,        .20),
+    "304":           (5.5,        .15),
+    "211":           (2.7,        .20),
+    "193":           (2.6,        .20),
+    "171":           (2.6,        .20),
+    "131":           (2.9,        .20),
+    "94":            (3.1,        .20),
+    "def":           (2.2,        .20)
+}
+
+
+def nafe(input_map, imargin=None, gamma=None, omargin=None, nafe=False,
+         nafe_weight=None, roi_p=None, roi_f=None, nproc=None, hbins=None,
+         nrs=15, gb=0.3, lb=0.7, gw=0.3, lw=0.7, n=129,
+         c=[1 for _ in range(12)]):
+    """
+    Enhances image "input_map" via "Noise Adaptive Fuzzy Equalization"
+    method for better visualization. Particularly suited for the exceptionally
+    high dynamic range images from the Atmospheric Imaging Assembly instrument
+    on the Solar Dynamics Observatory. This method produces artifact-free
+    images and gives significantly better results than methods based on
+    convolution or Fourier transform which are often used for that purpose.
+
+    Parameters
+    ----------
+    input_map : `~sunpy.map.sources.sdo.AIAMap`
+        The original input image to be enhanced.
+
+    imargin: `~tuple`
+        (black, white) input image values. Pixels with values < black or
+        > white are set to  0.
+
+    gamma : `~float`
+        The gamma value used for gamma_transform.
+
+    omargin: `~tuple`
+        (black, white) output image values.
+
+    nafe : `~bool`
+        Flag for enabling nafe transform and defaults to false.
+
+    nafe_weight : `~float`
+        Typical values for nafe_weight lie in the interval [0.05, 0.3].
+        Controls the ratio of the nafe component in the result image.
+
+    roi_p : `~tuple`
+        (row, col) of ROI top left corner pixel.
+
+    roi_f : `~tuple`
+        (width, height) of ROI frame.
+
+    nproc : `~int`
+        Number of processes for multiprocessing. Defaults to number of
+        available cpu cores.
+
+    hbins : `~int`
+        Number of histogram bins.
```

Comment:
NAFE in its core uses this histogram. No default mentioned in paper although it can be assumed to be 2^16 - 1 since this is the number of levels the paper uses. I thought if we are going to support images with even more levels to expose it as an argument. Currently default is calculated by: Freedman–Diaconis rule but I am sure something better can be done here.